src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -468,9 +623,56 @@ void ShenandoahFreeSet::recompute_bounds() {
   _mutator_leftmost = 0;
   _collector_rightmost = _max - 1;
   _collector_leftmost = 0;
+  _old_collector_rightmost = _max - 1;
+  _old_collector_leftmost = 0;
 
   // ...and adjust from there
   adjust_bounds();
+  if (_heap->mode()->is_generational()) {
+    size_t old_collector_middle = (_old_collector_leftmost + _old_collector_rightmost) / 2;
+    size_t old_collector_available_in_first_half = 0;
+    size_t old_collector_available_in_second_half = 0;
+
+    for (size_t index = _old_collector_leftmost; index < old_collector_middle; index++) {
+      if (is_old_collector_free(index)) {
+        ShenandoahHeapRegion* r = _heap->get_region(index);
+        old_collector_available_in_first_half += r->free();
+      }
+    }
+    for (size_t index = old_collector_middle; index <= _old_collector_rightmost; index++) {
+      if (is_old_collector_free(index)) {
+        ShenandoahHeapRegion* r = _heap->get_region(index);
+        old_collector_available_in_second_half += r->free();
+      }
+    }
+    _old_collector_search_left_to_right = (old_collector_available_in_second_half > old_collector_available_in_first_half);
+  }
+}
+
+bool ShenandoahFreeSet::expand_collector_bounds_maybe(size_t idx) {
```

Comment:
For this and the following method, the return value is not used, can be dropped.